pkg/capacity/capacity.go

```diff
@@ -0,0 +1,627 @@
+	// it, in case that it is already in the queue.
+	delete(c.capacities, item)
+
+	if capacity == nil {
+		// No object to remove.
+		klog.V(5).Infof("Capacity Controller: %+v removed, no object", item)
+		return
+	}
+
+	// Any capacity object in the queue will be deleted.
+	klog.V(5).Infof("Capacity Controller: enqueuing CSIStorageCapacity %s for removal", capacity.Name)
+	c.queue.Add(capacity)
+}
+
+// pollCapacities must be called periodically to detect when the underlying storage capacity has changed.
+func (c *Controller) pollCapacities() {
+	c.capacitiesLock.Lock()
+	defer c.capacitiesLock.Unlock()
+
+	for item := range c.capacities {
+		klog.V(5).Infof("Capacity Controller: enqueuing %+v for periodic update", item)
+		c.queue.Add(item)
+	}
+}
+
+func (c *Controller) runWorker(ctx context.Context) {
+	for c.processNextWorkItem(ctx) {
+	}
+}
+
+// processNextWorkItem processes items from queue.
+func (c *Controller) processNextWorkItem(ctx context.Context) bool {
+	obj, shutdown := c.queue.Get()
+	if shutdown {
+		return false
+	}
+
+	err := func() error {
+		defer c.queue.Done(obj)
+
+		switch obj := obj.(type) {
+		case workItem:
+			return c.syncCapacity(ctx, obj)
+		case *storagev1alpha1.CSIStorageCapacity:
+			return c.deleteCapacity(ctx, obj)
+		default:
+			klog.Warningf("unexpected work item %#v", obj)
+		}
+
+		return nil
+	}()
+
+	if err != nil {
+		utilruntime.HandleError(err)
+		klog.Warningf("Retrying %#v after %d failures", obj, c.queue.NumRequeues(obj))
+		c.queue.AddRateLimited(obj)
+	} else {
+		c.queue.Forget(obj)
+	}
+
+	return true
+}
+
+// syncCapacity gets the capacity and then updates or creates the object.
+func (c *Controller) syncCapacity(ctx context.Context, item workItem) error {
+	// We lock only while accessing c.capacities, but not during
+	// the potentially long-running operations. That is okay
+	// because there is only a single worker per item. In the
+	// unlikely case that the desired state of the item changes
+	// while we work on it, we will add or update an obsolete
+	// object which we then don't store and instead queue for
+	// removal.
+	c.capacitiesLock.Lock()
+	capacity, found := c.capacities[item]
+	c.capacitiesLock.Unlock()
+
+	klog.V(5).Infof("Capacity Controller: refreshing %+v", item)
+	if !found {
+		// The item was removed in the meantime. This can happen when the storage class
+		// or the topology segment are gone.
+		klog.V(5).Infof("Capacity Controller: %v became obsolete", item)
+		return nil
+	}
+
+	sc, err := c.scInformer.Lister().Get(item.storageClassName)
+	if err != nil {
+		if apierrs.IsNotFound(err) {
+			// Another indication that the value is no
+			// longer needed.
+			return nil
+		}
+		return fmt.Errorf("retrieve storage class for %+v: %v", item, err)
+	}
+
+	req := &csi.GetCapacityRequest{
+		Parameters: sc.Parameters,
+		// The assumption is that the capacity is independent of the
+		// capabilities. The standard makes it mandatory to pass something,
```

Comment:
Not sure. Perhaps it does make sense somewhere and it's just a simplifying assumption that is valid in Kubernetes, so let's gather feedback before proposing changes.

Reply:
Picking arbitrary values could get us in trouble if a plugin actually tries to validate any of them like they're supposed to. For now we're basically telling plugins "don't validate the capability for this call if you want it to work in Kubernetes". I think it would be a good to open an issue in the csi spec repo saying that when we tried to implement this, we could not find a meaningful use for this.

Reply:
Plugins can validate capabilities, because we are passing valid ones. What we are saying is "we expect capacity to be independent of this particular parameter". I created an issue for this: https://github.com/container-storage-interface/spec/issues/444